src/mahalanobis.jl

```diff
@@ -2,10 +2,24 @@
 
 struct Mahalanobis{M<:AbstractMatrix} <: Metric
     qmat::M
+    issemiposdef::Bool
```

Comment:
I still think there is no need to store this in a field and to adapt the algorithms since if it is not positive semi-definite then `Mahalanobis` and `SqMahalanobis` are not even `PreMetric`s.

In my opinion, one should just add a constructor that allows to skip the check, e.g., by defining
```julia
function Mahalanobis(Q::AbstractMatrix; issemiposdef=eigmin(Q) >= 0)
    ishermitian(Q) || throw(ArgumentError("bilinear form is not symmetric/Hermitian"))
    issemiposdef || throw(ArgumentError("bilinear form is not positive semidefinite"))
    return Mahalanobis{typeof(Q)}(Q)
end
```

Reply:
I thought there was a use case for indefinite matrices? https://github.com/JuliaStats/Distances.jl/issues/174#issuecomment-710688804 If we enforce both properties, then surely we don't need the field and the branches in the code.

Reply:
I think both comments mainly complained that the assumptions are not clearly enforced and hence it is easy to end up with nonsense results if one does not provide a proper positive semidefinite matrix. At least I don't have a use case for it and also the other comment mentioned that it is mainly an error on the user side.

Reply:
Then how about making it an optional argument `skipchecks`, make it default to `false` and skip both conditions if `skipchecks == true`? Then we avoid having to memorize how the argument is called.

Reply:
Ahh, that doesn't work because it overloads the standard constructor...

Reply:
Is there anything you dislike about the suggestion above? It seems reasonably explicit and direct and it would still allow one to skip all checks by just calling the inner constructor if desired.

Reply:
How do you call the inner constructor? `Mahalanobis(Q)` will append the default value for `issemiposdef`.

Reply:
`Mahalanobis{typeof(Q)}(Q)`.

Reply:
It's not that I dislike it, it's just that you have to remember the name of the kwarg, and I find it ugly to ask the user to type `Mahalanobis{typeof(Q)}(Q)` to avoid checks. What do you think of the approach in fa38714?

Reply:
I think it would be clearer with keyword arguments, it's not completely obvious what `true` in `Mahalanobis(M, true)` means.

Reply:
I'm fine with that. I moved the constructor to the inner part and made the argument a keyword argument.